packages/asset-swapper/src/utils/market_operation_utils/fills.ts

```diff
@@ -0,0 +1,305 @@
+): Fill[] {
+    // Create a single path from all orders.
+    let path: Fill[] = [];
+    for (const order of orders) {
+        const makerAmount = fillableAmountsUtils.getMakerAssetAmountSwappedAfterOrderFees(order);
+        const takerAmount = fillableAmountsUtils.getTakerAssetAmountSwappedAfterOrderFees(order);
+        const input = side === MarketOperation.Sell ? takerAmount : makerAmount;
+        const output = side === MarketOperation.Sell ? makerAmount : takerAmount;
+        const penalty = ethToOutputRate.times(fees[ERC20BridgeSource.Native] || 0);
+        const adjustedOutput = side === MarketOperation.Sell ? output.minus(penalty) : output.plus(penalty);
+        const rate = makerAmount.div(takerAmount);
+        const adjustedRate =
+            side === MarketOperation.Sell
+                ? makerAmount.minus(penalty).div(takerAmount)
+                : makerAmount.div(takerAmount.plus(penalty));
+        // Skip orders with rates that are <= 0.
+        if (adjustedRate.lte(0)) {
+            continue;
+        }
+        path.push({
+            input,
+            output,
+            rate,
+            adjustedRate,
+            adjustedOutput,
+            index: path.length,
+            flags: 0,
+            parent: path.length !== 0 ? path[path.length - 1] : undefined,
+            source: ERC20BridgeSource.Native,
+            fillData: { order },
+        });
+    }
+    // Sort by descending rate.
+    path = path.sort((a, b) => b.rate.comparedTo(a.rate));
+    // Re-index fills.
+    for (let i = 0; i < path.length; ++i) {
+        path[i].parent = i === 0 ? undefined : path[i - 1];
+        path[i].index = i;
+    }
+    return path;
+}
+
+function dexQuotesToPaths(
+    side: MarketOperation,
+    dexQuotes: DexSample[][],
+    ethToOutputRate: BigNumber,
+    fees: { [source: string]: BigNumber },
+): Fill[][] {
+    const paths: Fill[][] = [];
+    for (const quote of dexQuotes) {
+        const path: Fill[] = [];
+        for (let i = 0; i < quote.length; i++) {
+            const sample = quote[i];
+            const prevSample = i === 0 ? undefined : quote[i - 1];
+            const source = sample.source;
+            // Stop of the sample has zero output, which can occur if the source
+            // cannot fill the full amount.
+            // TODO(dorothy-zbornak): Sometimes Kyber will dip to zero then pick back up.
+            if (sample.output.eq(0)) {
+                break;
+            }
+            const input = sample.input.minus(prevSample ? prevSample.input : 0);
+            const output = sample.output.minus(prevSample ? prevSample.output : 0);
+            const penalty =
+                i === 0 // Only the first fill in a DEX path incurs a penalty.
+                    ? ethToOutputRate.times(fees[source] || 0)
+                    : ZERO_AMOUNT;
+            const adjustedOutput = side === MarketOperation.Sell ? output.minus(penalty) : output.plus(penalty);
```

Comment:
We may want to `break` if adjustedOutput is `lte(0)`, I guess optimizer would kick it out if there were other sources available.

Reply:
I considered this at first but realized it could drop entire paths because the fee in the first fill can drive it to negative (esp if we increase samples). This might be viable for native paths, though.

Reply:
Oh we already do that for native paths.

Reply:
ah I see, that makes sense, especially for small fills and 20 samples.